### ISSUE TITLE
Pre-trained model

### ISSUE THREAD

**Issue by JakubBartoszewicz** [KennthShang/PhaTYP#5]
Hi,

Thanks for developing the tool! Is the pre-trained model (without any fine-tuning, but the same that was used for fine-tuning later) available? Could you maybe also share it?

Thank you!

**Comment by KennthShang** [comment]
The model had already been uploaded. You only need to follow the guidelines to download and install.

Please check whether you have followed the commands below:

```
cd PhaTYP/
conda env create -f phatyp.yaml -n phatyp
conda activate phatyp
cd database/
bzip2 -d database.fa.bz2
cd ..

fileid="1tsUArctGf9Fd3xa-0sEcp6ykwxTy9uxG"
filename="model.zip"
html=`curl -c ./cookie -s -L "https://drive.google.com/uc?export=download&id=${fileid}"`
curl -Lb ./cookie "https://drive.google.com/uc?export=download&`echo ${html}|grep -Po '(confirm=[a-zA-Z0-9\-_]+)'`&id=${fileid}" -o ${filename}

unzip model.zip

```

**Comment by JakubBartoszewicz** [comment]
Hi @KennthShang, thanks for the quick reply. However, this doesn't look like what we need - I've of course read the readme, but I thought this is the final, fine-tuned model that you run predictions with, right? We need something else - only the BERT model, without fine-tuning, basically the output of `pretrain.py` and input for your `finetune.py` script. We'd like to do fine-tuning ourselves, without any data leakage between the fine-tuning datasets that you used and those that we will. For that, we'd like to use the *exact* same pre-trained model as you used as input to `finetune.py` when generating results in your Table 4 in your paper. Thanks!

**Comment by KennthShang** [comment]
sorry that i do have my disk with me in lyon. will check after back to hk

**Comment by KennthShang** [comment]
but i suppose there might exist other problem if you train on deephage and has 50% accuracy. you can report it in your paper if you are in a hurry. Its ok. I will try to figure out it in the near future.

**Comment by KennthShang** [comment]
Hi, there,

I have several questions about the thing you told me today. 

1. How did you fine-tune the model? Because according to our experiences, if we enlarge the epoch for the model-only training on phage lifestyle prediction with MaskLM pretraining (for example 100X than the previous one), its performance can be the same as the model. So, if you have an accuracy of about 50%, that is not caused by the pertaining module. I also ran a quick test with the dataset provided by DeePhage, and the results look good. I suppose you have a problem constructing the dataset using our scripts.
2. How did you generate your training and test data? Because the input format for Bert is special, you still need to add the labels for the input files as mentioned in the guidelines. Also, the script can only generate input files for virulent and temperate phages independently. If you combine the files and run the preprocessing.py, problems might exist. This is because some of the phages cannot be converted into sentences due to the alignment results. So if you combine virulent and temperate phages into one file and call the preprocessing function, your index might differ greatly. 
3. BTW, I am not sure how you use the data from DeePhage because some of the phages in DeePhage data have the wrong label. This can be easily checked by using the BLAST search. I removed these sequences when training our final models. In your experiment, you could construct the dataset on your own.
4. It will be ok if PhaTYP cannot do some of the cases in your experiments. There might exist some situations we may not consider in our experiments. I am glad you can share this information with me and willing to read your paper in the future.

Best,
Jiayu

**Comment by JakubBartoszewicz** [comment]
Hi,

Thank for coming back to us and sorry for a delayed reply - Lyon was busy ;) and we're currently out of office. Everything below is in the context of short contigs.

1. We finetuned it like that: `torchrun --nproc_per_node=2 finetune.py --train ../train.csv --val ../test.csv`. Since you ran the checks - does this mean you do have the original pretrained (non-finetuned) model available? Could you share it? What was the contig length you used for the check? How exactly did you split the data into training and test?
2. The virulent and temperate classes were preprocessed separately using your `preprocessing.py` script for all runs we did. We then set the labels for both classes accordingly (0 or 1 in the `label` column, respectively), as in your `finetune_example.csv` file.
3. We did two things:
- *A)* we used the exact data that you shared in your repo, `dataset320000`, contig length 100-400. The exact training-test splits you used were not available, so we generated a 90%-10% split ourselves (as the first fold of your 10-fold CV; but we didn't continue given the problems). What do you mean that you removed some of the sequences? Your repo seems to say this is the exact dataset you used. These files are *identical* to the ones found in the DeePhage SI in `all_data_to_train_model/100_400`. But you write in the paper that you generated those yourself? Or are those not the final files, and you still do the Dashing step on them, and train only afterwards? Is the removal of 'wrong' labels described somewhere in your paper? Sorry if I missed it.
- *B)* The results for DeePhage, PhagePred & PHACTS in Table 4 of your paper seem to be identical with the results in Table 1 of the DeePhage paper. However, they used a 5-fold CV where *different phages* were placed in each fold. The files with simulated contigs split by fold are in the DeePhage SI in `training/MetaSim_simulating/`. Therefore, we tried finetuning PhaTYP on the exact same folds that were used to generate the results in the tables. This didn't work either.

By the way, you wrote in your paper that you did 10-fold CV, and it seems that you used the files in `all_data_to_train_model/100_400` as input, right? This would mean that different contigs of the same phage can be in both training and test set of any given fold, right? Did you also run a similar 10-fold CV on the same data for DeePhage, PhagePred & PHACTS? It's surprising that the results are the same as for a more strict 5-fold CV (where contigs of the same phage cannot be in training and test) presented in the DeePhage paper.

In any case, you're right, there must be a bug somewhere in the process, I can't see how none of those experiments would work... Sharing the exact pre-trained model, code and/or the exact data you use would help with debugging. I was wondering if the problem could be that our reproduction of your pre-training failed. The loss does not seem to change much over pre-training. What is the eval loss value you'd expect?

Thanks & best,
Jakub

**Comment by KennthShang** [comment]
I think you may misunderstand what we have in the paper.

1. The data of the short contigs experiments we used in the paper are provided by the DeePhage. The short contigs we provided in the google drive are used to train the final model we uploaded to this GitHub. In the paper's experiments, all the benchmarked datasets (expected the one in the section *Performance on the test set with low similarity*) are from the DeePhage, and we follow the same rules to train and evaluate PhaTYP as DeePhage. Also, in each section, we re-train the PhaTYP model using the fine-tuning method as DeePhage.
2. The dataset used in section *Performance on the test set with low similarity* was generated by ourselves. However, it is still the same data provided by DeePhage. What we did was use Dashing to estimate the all-against-all similarity within the dataset and split them accordingly. The similarity files can be found via https://drive.google.com/file/d/1dDaYuk7uscnlLzMvee8NDkS-tys4K9SU/view?usp=sharing
3. This is not a problem of the pre-training but a problem of fine-tuning. As we reported in the paper, we did an ablation experiment. The results of pre-training are about ~2% improvement from the ones directly training on the lifestyle data. So what does matter is the fine-tuning progress. But, of course, the pre-trained model can be downloaded via: https://drive.google.com/file/d/1zzrBaV5dPTljdhRncFax34T696tm8TQQ/view?usp=sharing
4. For your convenience, I also provide an example file for a quick test. You can use the fine-tuned script and the pre-trained model on this file (both train and test, just to test whether the script works). After training with 5 epochs, the accuracy will be about 97%. The example file can be found via:https://drive.google.com/file/d/1l_kmqH10juul3BUBxYDam51MNRlmLAEe/view?usp=sharing
5. The removed phages mean: You can use the marker genes, such as integrase, to align the data provided by DeePhage. Then you will find that some of the virulent phages can be aligned to these proteins with 100% identity. So I remove them in the final model.

Hope all the information above will help your work. BTW, PhaTYP may not always perform well; if you are testing on your data, we are not sure what will happen. But we will consistently update our model when there exists a more reliable dataset for biologists and ecologists. So the model provided by PhaTYP will change time by time.

**Comment by JakubBartoszewicz** [comment]
Hi,

Thank you very much. I'm afraid the description in the paper might be less clear than you'd like it to be then... Apologies for being very direct and technical below - I don't intend to come off as rude, but since you suggest it's a misunderstanding, let's try to avoid any miscommunication and any lack of clarity (especially since we run our experiments over half a year ago, and we also understand remembering all the details of your training procedures on the spot might be difficult). If anything remains unclear, let me know and I will try to explain it better. Briefly, this is not exactly about testing on our data here (of course, we know your method will not work for anything anytime), but about reproducibility of the presented results, which requires exact knowledge of what the training and test sets were for each experiment and CV fold.

- In the [readme](https://github.com/KennthShang/PhaTYP/blob/main/train/README.md), you write that _"All the contigs dataset used in the paper can be downloaded via google drive"_ and that they represent _"the three sets of data used in the experiments"_. Therefore, I assume these three datasets (dataset142434, dataset160000, dataset320000) are the only ones that have been used and allow at least some level of replication of all the experiments from the paper. Is this correct?
- In _Results_, section _Performance on the short contigs_, paragaph 1, you write: _"**we generate** four groups of fragments with different length ranges"_. However, it seems those are the contigs generated by the DeePhage authors, provided by them, as you mentioned yourself above. Is this correct?
- In the same paragraph, you write: _"We applied ten-fold cross-validation to train and validate the performance."_, later referring to Table 4. You also say that this was done using the same "rules"/procedure as DeePhage did for an analogous benchmark (their Table 1, with identical results for the other methods). However, they used five-fold cross-validation. These are not the same _"rules"_. Since they used 5-fold, and you used 10-fold, at least the splitting procedure must have been different. Is this correct?
- In the DeePhage paper, section _Results_, first paragraph, the authors write that they performed the training-validation split for their 5-fold CV _"on the basis of complete genomes rather than artificial contigs"_. We were able to reproduce their results using the 5 folds provided by them with no problems. Were the splits of your 10-fold CV also done by assigning whole genomes to train/test first, and simulating the contigs later, as they have apparently done it? If so, are these splits available?
- In the same paragraph in section _Performance on the short contigs_ , you mention using Dashing to remove fragments with similarity above 0.8. dataset320000 is identical to the original one provided by DeePhage, with no trace of removal of any sequences. I do not mean the section _Performance on the test set with low similarity_ here. This step is also a deviation from the DeePhage "rules"/procedure (although it does make a lot of sense). I assume it was indeed done here, but was it done before the 10-fold split (as the order of sentence in your paper suggests) or after it (if you followed the DeePhage procedure of splitting by genome first, and simulating contigs later)?

> This is not a problem of the pre-training but a problem of fine-tuning.

I believe you seem to be wrong here, but it also seems to be good news after all! I've just used the pretrained model shared by you above (using the exact same setup as before with the model pre-trained by us) and it works now, at least in a quick test with a simple 9-1 split of dataset320000. Thanks, that was extremely helpful! This would confirm my previous suspicion that our replication of the pre-training step didn't really work. Not sure why though.

**Comment by KennthShang** [comment]
Hmm, I think I made a mistake or an unclear description for the readme file haha. Very sorry about that. The data from google drive is only the final model for usage proper. It contains all the data we have and we generate much more fragments to help the model to be stable (because Bert really has much more params).

Also, I am afraid that I do misremembered the detail of the paper. The work was done in 2021, but it has a long-term revision during the submission process. So I check the workflow and the revision letter. The first version was done using the same data (short fragments) provided by DeePhage. But then, the reviewer said, we need to split them into more folds rather than five. So we repeat it again as the rules in DeePhage.

However, I did not have the dataset remaining here. So maybe you can try to generate them.

BTW, it will be strange if the bert cannot work without pre-training. Maybe the training epoch matters. But I am glad to solve your problems.

Have a nice day.

Best,
Jiayu
